src/Domain/ElementDistribution.hpp

```diff
@@ -0,0 +1,92 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
+#include "Domain/Structure/ElementId.hpp"
+
+namespace domain {
+
+/*!
+ * \brief Distribution strategy based on grouping elements within blocks
+ * together, and using a Morton ('Z-order') space-filling curve to determine
+ * distribution within blocks.
+ *
+ * \details This class suggests an element placement based on a Z-curve/Morton
+ * space filling curves.
+ * Morton curves are a simple and easily-computed space-filling curve that
+ * (unlike Hilbert curves) permit diagonal traversal. See, for instance,
+ * \cite Borrell2018 for a discussion of mesh partitioning using space-filling
+ * curves.
+ *
+ * First, a curve is computed through each block. Then the elements are assigned
+ * to a processor in chunks such that each processor has approximately the same
+ * number of elements. This is done by starting with block zero, assigning the
+ * first chunk to processor 0 on node 0, the second to processor 1 on node 0,
+ * and so on until node 0 is filled. The process then repeats on the next node.
+ * If the number of elements in a block is not a multiple of the element chunk
+ * size, then part of the next block is also added to the processor. For
+ * example, processor 1 may receive half its elements from block 0 and half
+ * from block 1. A concrete example in 2d is given below.
```

Comment:
Right, I admit this can be improved, but we'll have to work on it. The trouble is that we actually don't have any global curve -- Here's another attempt:

First, we perform a greedy assignment of a balanced number of elements to each processor by giving each processor an allowance of [total number of elements]/[number of processors] elements from one or more blocks, starting with the lowest number block that still has elements to contribute to an allowance.
Then, once those allowances are determined, a separate z-order curve is established for each block, and the elements are assigned to processors within each block by greedily filling each processors' allowance in contiguous intervals along the z-curve.

Does that make more sense? Or is the use of the term 'allowance' just more confusing?

Reply:
That does make more sense. Let's check my understanding:

1. 100 elements across 10 blocks on 10 CPUs => the "allowance" is 10 elements/CPU, each CPU handles the 10 elements from 1 block
2. 100 elements across 5 blocks on 10 CPUs => allowance is 10 again, each CPU handles 10 elements from half a block
3. 100 elements across 20 blocks on 10 CPUs => allowance still 10, each CPU handles 10 elements in two "chunks" of 5, each chunk coming from 1 block
4. 100 elements across 10 blocks on 8 CPUs => the allowance is 12.5 elements/CPU, so CPU0 has 10 elements from block0 + 2(ish) elements from block1, then CPU1 has 8(ish) elements from block1 and 4(ish) elements from block2, etc.

Is this more correct?

Reply:
Yep, that's right -- it's not perfect because it doesn't at all consider the geometric relationship between the blocks, but whenever the blocks have a significant number of elements, I think the distribution should end up being in fairly large clusters.

Reply:
👍 Then I think your new proposed text is good. I'd suggest adding some examples (like what I was checking above, but maybe with better numbers to avoid the fractional allowance (powers of 2 perhaps?)) to help the reader check their understanding.. what do you think?

Reply:
Sounds good -- I added another fixup commit with a couple of examples of this 'allowance' logic.